Add flushing support to fasthttpadaptor

grivera64 opened the pull request:
### Description

Currently, `fasthttpadaptor` runs the net/http handler fully first, then copies the response to a fasthttp response object and returns. This makes it incompatible with net/http handlers that may rely on instantly flushing to the client (e.g. Live Data Streams, SSE events).

This PR adds support flushing by concurrently handling the net/http handler and copy chunks of data to the fasthttp response after a flush occurs. If the response is never flushed, the `fasthttpadaptor` can follow its old behavior and send the data as a single chunk for efficiency.

The PR also adds a few new test cases to ensure that flushed data is received if the client connection closes before the full response is sent.

Fixes: #2053 
Related: https://github.com/gofiber/fiber/issues/3421

### Additional Comments

This feature will help close a major gap in fasthttp's `net/http` support, which is a main barrier of adoption for `fasthttp`. Flushing support would enable more `net/http` handlers to work with minimal overhead.

erikdubbelboer commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,42 +56,169 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
-				l = len(b)
+
+			if len(w.firstChunk) > 0 {
+				ctx.Response.SetBodyRaw(append([]byte(nil), w.firstChunk...))
```
Why the append with `nil` here?

grivera64 replied:
I've changed it to now just directly set the raw body to `w.responseBody`.

erikdubbelboer commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,42 +56,169 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+	flushedCh   chan struct{}
+	streamingCh chan struct{}
+	hijackedCh  chan struct{}
+	firstChunk  []byte
```
`firstChunk` is a bit of a weird name as it's only the first chunk in case of streaming, for the non-streaming case it's the full response. Maybe just call it something like `responseBody`.

erikdubbelboer reviewed (changes requested):
I think there's a race condition. What if the handler calls Flush and right after that calls Write. That Write will then race with `<-w.flushedCh` case on who gets to access `firstChunk` first.

There also seems to be a race condition with the statusCode, but that is less likely to happen as it doesn't make sense to call `WriteHeader` after `Flush`.

`w.r`, the reading side of the Pipe, is never closed. I'm not sure if this will lead to leaks.

The whole thing causes a lot more heap allocations, but I guess if you're using a net/http handler with fasthttp you don't really care about performance anyways.

grivera64 commented:
> I think there's a race condition. What if the handler calls Flush and right after that calls Write. That Write will then race with `<-w.flushedCh` case on who gets to access `firstChunk` first.
> 
> There also seems to be a race condition with the statusCode, but that is less likely to happen as it doesn't make sense to call `WriteHeader` after `Flush`.

I've applied the requested changes, and added two mutexes to help with the race conditions.

> `w.r`, the reading side of the Pipe, is never closed. I'm not sure if this will lead to leaks.

I now added closing the read pipe to the `Close()` method.

> The whole thing causes a lot more heap allocations, but I guess if you're using a net/http handler with fasthttp you don't really care about performance anyways.

It should still be largely recommended to port `net/http` to `fasthttp` for maximum performance. This adapter is more of an alternative for users that want to use a `net/http` handler in a `fasthttp` server, but do not currently have the time to port the code yet and don't mind the overhead.

grivera64 commented:
@erikdubbelboer I applied one of the copilot hints and added a `sync.Pool` to help reduce memory allocations whenever acquiring a `netHTTPResponseWriter` object. This should help reduce the memory overhead of serving a `net/http` handler.

grivera64 commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,104 +57,304 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+func acquireNetHTTPResponseWriter(ctx *fasthttp.RequestCtx) *netHTTPResponseWriter {
+	w, ok := writerPool.Get().(*netHTTPResponseWriter)
+	if !ok {
+		log.Panic("cannot get from writerPool")
```
I've changed it to use a normal `panic` with the improved message from the hint above.

grivera64 commented:
@erikdubbelboer The race conditions should be gone now. I ran `go run gotest.tools/gotestsum@latest -f testname -- ./.fasthttpadaptor -race -count=15 -shuffle=on` to check for rare race conditions, but all runs succeed.

grivera64 commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,104 +56,339 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+			}()
+			wg.Wait()
+
+			// Wait for the net/http handler to finish
+			// writing to the hijacked connection prior to releasing
+			// the writer into the writer pool.
+			<-doneCh
```
I am unsure if waiting for the `doneCh` is necessary here. It seemed to help with race conditions before, but `wg.Wait()` in theory should wait until the `w.handlerConn` is closed. This only happens after the handler returns.

I've commented it out locally, and tests still pass. It may be a good idea to remove it once confirmed to be unnecessary.

grivera64 replied:
It seems that the `<-doneCh` logic isn't correct to include here. I have removed it to avoid infinitely blocking the release of the writer here.

grivera64 commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,104 +56,339 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+					// Copy chunk to fasthttp response
+					if n > 0 {
+						_, _ = bw.Write((*chunk)[:n])
+						_ = bw.Flush()
```
I've fixed this by checking if the `bw.Write()` or `bw.Flush()` fail. If any of those two methods fail, then it will break out of the loop, avoiding an infinite loop.

grivera64 commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,104 +56,369 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+}
+
+func acquireBuffer() *[]byte {
+	buf, ok := bufferPool.Get().(*[]byte)
+	if !ok {
+		panic("fasthttpadaptor: cannot get *[]byte from bufferPool")
+	}
```
This is okay, since this will ensure that future tests will fail upon modifying the `bufferPool`.

grivera64 commented on fasthttpadaptor/adaptor.go:
```diff
@@ -56,104 +56,369 @@ func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
+	if w.handlerConn != nil {
+		w.handlerConn.Close()
+	}
+	w.connMutex.Unlock()
```
This has been updated to explicitly ignore the `handlerConn.Close()` error.

grivera64 commented:
@erikdubbelboer I was able to remove most allocations per request by removing most of the channels in favor of using a single channel, `sync.Cond`, and `sync.Once`. The test cases pass, and the code works as expected for the following `net/http` handlers:

<details><summary>Flush net/http Handler (Click to Open)</summary>
<p>

```go
package main

import (
	"bufio"
	"fmt"
	"io"

	"net/http"
	"time"

	"github.com/valyala/fasthttp"
	"github.com/valyala/fasthttp/fasthttpadaptor"
)

func main() {
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.Header().Set("Cache-Control", "no-cache")

		bw := bufio.NewWriter(w)
		defer bw.Flush()

		for i := 0; i <= 10; i++ {
			msg := fmt.Sprintf("Line %d\n", i)
			_, err := bw.WriteString(msg)
			if err != nil {
				if err == io.ErrClosedPipe {
					return
				}
				fmt.Println("net/http: write error:", err)
				return
			}
			err = bw.Flush()
			if err != nil {
				if err == io.ErrClosedPipe {
					return
				}
				fmt.Println("net/http: flush error:", err)
				return
			}
			if flusher, ok := w.(http.Flusher); ok {
				flusher.Flush()
			}
			time.Sleep(1 * time.Second)
		}
	})

	fastHTTPHandler := fasthttpadaptor.NewFastHTTPHandler(handler)
	fasthttp.ListenAndServe(":3000", fastHTTPHandler)
}
```

</p>
</details> 

<details><summary>Hijack net/http handler (Click to Open)</summary>
<p>

```go
package main

import (
	"fmt"

	"log"
	"net/http"
	"time"

	"github.com/valyala/fasthttp"
	"github.com/valyala/fasthttp/fasthttpadaptor"
)

func main() {
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		hj, ok := w.(http.Hijacker)
		if !ok {
			http.Error(w, "webserver doesn't support hijacking", http.StatusInternalServerError)
			return
		}

		conn, rw, err := hj.Hijack()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		defer conn.Close()

		_, err = rw.WriteString("HTTP/1.1 200 OK\r\n" +
			"Content-Type: text/plain; charset=utf-8\r\n" +
			"Cache-Control: no-cache\r\n" +
			"\r\n")
		if err != nil {
			log.Printf("Error writing initial headers: %v", err)
			return
		}
		rw.Flush()

		for i := 0; i <= 10; i++ {
			msg := fmt.Sprintf("Line %d\n", i)
			_, err := rw.WriteString(msg)
			if err != nil {
				log.Printf("Error writing to connection: %v", err)
				return
			}

			err = rw.Flush()
			if err != nil {
				log.Printf("Error flushing to connection: %v", err)
				return
			}

			time.Sleep(1 * time.Second)
		}
	})

	fastHTTPHandler := fasthttpadaptor.NewFastHTTPHandler(handler)
	fasthttp.ListenAndServe(":3000", fastHTTPHandler)
}
```

</p>
</details>

The memory allocations should be at a minimum, though I haven't formally run `pprof` against the implementation. Please let me know if you have any thoughts on the new implementation.

grivera64 commented:
Fixed the Lint issues. For some reason, `golangci-lint` doesn't work for me locally. The lint workflow should pass now.

Aside from that, the other workflows seem to pass (aside from an unrelated error in `TestDirFSFSCompressConcurrent`).
